### PR TITLE
Validate the existence of the model specified in --model

### LINKED ISSUE #753: In 'lab chat', tell the user if you are using a default model, please validate --model or -m before running
After running lab init, lab check, lab generate, lab train --device cuda (on Linux), I end up with a new .gguf file:

```bash
$ ls -l models/
total 14145392
-rw-rw-r-- 1 dperique dperique 14484862880 Mar 25 16:43 ggml-model-f16.gguf
lrwxrwxrwx 1 dperique dperique         148 Mar 15 20:56 merlinite-7b-Q4_K_M.gguf -> ../../../../../../../../.cache/huggingface/hub/models--ibm--merlinite-7b-GGUF/blobs/94f3a16321c9604ca22e970f3b89931ae5b4bbfd4c5d996e2bb606c506590666
```

This is the merged model (i.e., original model plus adapter).  If I try to chat with it via `lab chat -m ggml-model-f16.gguf`, it appears to load:

```bash
$ lab chat -m ggml-model-f16.gguf 
╭──────────────────────────────────────────────────────────── system ────────────────────────────────────────────────────────────╮
│ Welcome to Chat CLI w/ GGML-MODEL-F16.GGUF (type /h for help)                                                                  │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
>>>                                                                                                                   [S][default]
```

But that is NOT the correct model -- even though it confidently implies it in the "Welcome ..." message.

You can observe this more clearly using a non existent file:

```bash
$ lab chat -m something_non_existent.lala
╭──────────────────────────────────────────────────────────── system ────────────────────────────────────────────────────────────╮
│ Welcome to Chat CLI w/ SOMETHING_NON_EXISTENT.LALA (type /h for help)                                                          │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
>>>            
```

Clearly, we're defaulting to something (probably in the `config.yaml`) but the user should be informed explicitly -- silence is not acceptable.

With silence, the user may try to test the model manually only to find it does not appear to be fine-tuned.  This in fact was my case as I spent a lot of time trying to understand why my resultant gguf file didn't seem to incorporate my dataset.

Using the correct file loads the expected model:

```bash
$ lab chat -m ./models/ggml-model-f16.gguf
╭──────────────────────────────────────────────────────────── system ────────────────────────────────────────────────────────────╮
│ Welcome to Chat CLI w/ ./MODELS/GGML-MODEL-F16.GGUF (type /h for help)                                                         │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
>>> 
```

And my tests show the model was fine-tuned as expected.

We should validate the existence of the value of the `--model` or `-m` parameter to avoid the user being confused about what model is being used and/or invalidating long hours of testing.

### PR DESCRIPTION
**Pull request by cdoern**
# Changes

the openAI client that connects to endpoint_url has a list of models the server can chat with.

We need to look through that list and ensure the user specified --model or -m exists in that list

**Which issue is resolved by this Pull Request:**
Resolves #753

### CODE REVIEW

**Review by russellb — CHANGES_REQUESTED** [review]
Let's hold this until after #935 since they are so closely related.

Also - this change touches `chat`, but the same issue you're addressing here applies to `generate` as well.

**Comment by russellb** [comment]
I removed the hold because #935 was merged. This will have to be updated to account for those changes. We should definitely validate the requested model and raise a friendly error if it's not there, but you'll have to account for the code from #935 that handles the case of running a non-default model from `ilab serve` and then running `ilab chat` with no arguments. That has always worked even though the code asks for the wrong model name. #935 made it so the chat UI reflected the actual model we connected to.

When something specific and non-default is requested though, whether it's local or a remote endpoint URL, this validation is good and can help us give a more friendly error than getting a 404 Not Found error a bit later.

**Comment by nathan-weinberg** [comment]
@Mergifyio rebase

File: src/instructlab/chat/chat.py
Comment (leseb) [inline src/instructlab/chat/chat.py:488]:
```suggestion
        raise ChatException(f"Model path {model} does not exist on the server.")
```

Also, I'd print the list of available models to help the user further. Thanks!

**Review by leseb — COMMENTED** [review]
I believe this is getting close, thanks for your patience.

File: src/instructlab/chat/chat.py
Comment (hickeyma) [reply src/instructlab/chat/chat.py:488]: Might read better as: "Model {model} is not served by the server."?

**Review by hickeyma — CHANGES_REQUESTED** [review]
Nice check, thanks for adding @cdoern. One small nit.

**Review by nathan-weinberg — APPROVED** [review]

**Review by russellb — CHANGES_REQUESTED** [review]
I think this comment still applies?  let me know if I missed something

> Also - this change touches chat, but the same issue you're addressing here applies to generate as well.

**Comment by cdoern** [comment]
sorry @russellb I totally missed that! Fixed now, want to let the tests run since I realized running `ilab chat` fails since the default for the model flag is now incorrect

**Review by russellb — APPROVED** [review]
lgtm pending CI

**Review by russellb — CHANGES_REQUESTED** [review]
spoke too soon, CI failed

>   Executing chat failed with: Model merlinite-7b-lab-Q4_K_M is not served by the server.

**Comment by cdoern** [comment]
@russellb I did the following and it worked:

1. nano config.yaml, change it to: 
```chat:
  context: default
  greedy_mode: false
  logs_dir: data/chatlogs
  max_tokens: null
  model: models/merlinite-7b-lab-Q4_K_M.gguf
  session: null
  vi_mode: false
  visible_overflow: true
general:
  log_level: INFO
generate:
  chunk_word_count: 1000
  model: models/merlinite-7b-lab-Q4_K_M.gguf
  num_cpus: 10
  num_instructions: 100
  output_dir: generated
  prompt_file: prompt.txt
  seed_file: seed_tasks.json
  taxonomy_base: origin/main
  taxonomy_path: taxonomy
serve:
  gpu_layers: -1
  host_port: 127.0.0.1:8000
  max_ctx_size: 4096
  model_path: models/granite-7b-lab-Q4_K_M.gguf
```
2. ilab serve
3. ilab chat
```
(venv) cdoern@cdoern-thinkpadp16vgen1:~/Documents/instructlab$ ilab chat
╭────────────────────────────────────────────── system ───────────────────────────────────────────────╮
│ Welcome to InstructLab Chat w/ MODELS/GRANITE-7B-LAB-Q4_K_M.GGUF (type /h for help)                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
>>>                                                                                        [S][default]

```

let me know if you are doing something different

**Comment by cdoern** [comment]
maybe you didn't specify the .gguf part?

File: scripts/functional-tests.sh
Comment (leseb) [inline scripts/functional-tests.sh:347]: Why is this being removed?

File: src/instructlab/config.py
Comment (hickeyma) [inline src/instructlab/config.py:23]: Why are we making changes here? Is it required for the PR goal?

File: scripts/functional-tests.sh
Comment (hickeyma) [reply scripts/functional-tests.sh:347]: Agree

**Review by hickeyma — DISMISSED** [review]
@cdoern Can you explain changes that are not directly related to validating the existence of the model?

File: scripts/functional-tests.sh
Comment (cdoern) [reply scripts/functional-tests.sh:347]: this is testing whether the user can put something like baz.gguf, this should not work and will not. The new code only allows  -m which is in the server

File: src/instructlab/config.py
Comment (cdoern) [reply src/instructlab/config.py:23]: The default model needs to match the full path. The -m the user gives and/or the default need to be the fully qualified path to the served model, especially in vllm.

**Comment by cdoern** [comment]
@hickeyma @russellb PTAL, I would like this fix to be in for today's release

File: src/instructlab/config.py
Comment (cdoern) [reply src/instructlab/config.py:23]: something like `merlinite-7b-lab` isn't ever going to be one of the models found when querying the server, only the full path will be recognized

File: scripts/functional-tests.sh
Comment (cdoern) [reply scripts/functional-tests.sh:347]: `ilab chat -m bar` should really never have worked, this PR makes it so that -m only works for real, fully qualified models.

File: src/instructlab/config.py
Comment (leseb) [inline src/instructlab/config.py:23]: Since both `DEFAULT_MODEL` and `DEFAULT_MODEL_PATH` share the same value. Why keep both?

**Review by leseb — COMMENTED** [review]
Small nit question.

File: scripts/functional-tests.sh
Comment (cdoern) [reply scripts/functional-tests.sh:347]: sure

File: src/instructlab/config.py
Comment (cdoern) [reply src/instructlab/config.py:23]: Adding a todo, I think this  consolidation is out of scope for this PR as that would involve changing cmd flags since these vars auto populate those, right?

```# TODO: Consolidate --model and --model-path into one --model-path flag since we always need a path now
```

**Comment by russellb** [comment]
> @russellb I did the following and it worked:
> 
> 1. nano config.yaml, change it to:
> 
> ```
>   context: default
>   greedy_mode: false
>   logs_dir: data/chatlogs
>   max_tokens: null
>   model: models/merlinite-7b-lab-Q4_K_M.gguf
>   session: null
>   vi_mode: false
>   visible_overflow: true
> general:
>   log_level: INFO
> generate:
>   chunk_word_count: 1000
>   model: models/merlinite-7b-lab-Q4_K_M.gguf
>   num_cpus: 10
>   num_instructions: 100
>   output_dir: generated
>   prompt_file: prompt.txt
>   seed_file: seed_tasks.json
>   taxonomy_base: origin/main
>   taxonomy_path: taxonomy
> serve:
>   gpu_layers: -1
>   host_port: 127.0.0.1:8000
>   max_ctx_size: 4096
>   model_path: models/granite-7b-lab-Q4_K_M.gguf
> ```
> 
> 2. ilab serve
> 3. ilab chat
> 
> ```
> (venv) cdoern@cdoern-thinkpadp16vgen1:~/Documents/instructlab$ ilab chat
> ╭────────────────────────────────────────────── system ───────────────────────────────────────────────╮
> │ Welcome to InstructLab Chat w/ MODELS/GRANITE-7B-LAB-Q4_K_M.GGUF (type /h for help)                 │
> ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
> >>>                                                                                        [S][default]
> ```
> 
> let me know if you are doing something different

It looks like the difference is that your `config.yaml` was generated after the changes made in your PR. If you have a config using the defaults from prior to this PR, it breaks.

```
chat:
  context: default
  greedy_mode: false
  logs_dir: data/chatlogs
  max_tokens: null
  model: merlinite-7b-lab-Q4_K_M
  session: null
  vi_mode: false
  visible_overflow: true
general:
  log_level: INFO
generate:
  chunk_word_count: 1000
  model: merlinite-7b-lab-Q4_K_M
  num_cpus: 10
  num_instructions: 100
  output_dir: generated
  prompt_file: prompt.txt
  seed_file: seed_tasks.json
  taxonomy_base: origin/main
  taxonomy_path: taxonomy
serve:
  gpu_layers: -1
  host_port: 127.0.0.1:8000
  max_ctx_size: 4096
  model_path: models/merlinite-7b-lab-Q4_K_M.gguf
```

```
$ ilab serve
$ ilab chat
Executing chat failed with: Model merlinite-7b-lab-Q4_K_M is not served by the server.

File: src/instructlab/config.py
Comment (leseb) [reply src/instructlab/config.py:24]: I'm fine with that, another small request, could you please open an issue so we can track it more easily? Thanks!

**Comment by cdoern** [comment]
> > @russellb I did the following and it worked:
> > 
> > 1. nano config.yaml, change it to:
> > 
> > ```
> >   context: default
> >   greedy_mode: false
> >   logs_dir: data/chatlogs
> >   max_tokens: null
> >   model: models/merlinite-7b-lab-Q4_K_M.gguf
> >   session: null
> >   vi_mode: false
> >   visible_overflow: true
> > general:
> >   log_level: INFO
> > generate:
> >   chunk_word_count: 1000
> >   model: models/merlinite-7b-lab-Q4_K_M.gguf
> >   num_cpus: 10
> >   num_instructions: 100
> >   output_dir: generated
> >   prompt_file: prompt.txt
> >   seed_file: seed_tasks.json
> >   taxonomy_base: origin/main
> >   taxonomy_path: taxonomy
> > serve:
> >   gpu_layers: -1
> >   host_port: 127.0.0.1:8000
> >   max_ctx_size: 4096
> >   model_path: models/granite-7b-lab-Q4_K_M.gguf
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. ilab serve
> > 3. ilab chat
> > 
> > ```
> > (venv) cdoern@cdoern-thinkpadp16vgen1:~/Documents/instructlab$ ilab chat
> > ╭────────────────────────────────────────────── system ───────────────────────────────────────────────╮
> > │ Welcome to InstructLab Chat w/ MODELS/GRANITE-7B-LAB-Q4_K_M.GGUF (type /h for help)                 │
> > ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯
> > >>>                                                                                        [S][default]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > let me know if you are doing something different
> 
> It looks like the difference is that your `config.yaml` was generated after the changes made in your PR. If you have a config using the defaults from prior to this PR, it breaks.
> 
> ```
> chat:
>   context: default
>   greedy_mode: false
>   logs_dir: data/chatlogs
>   max_tokens: null
>   model: merlinite-7b-lab-Q4_K_M
>   session: null
>   vi_mode: false
>   visible_overflow: true
> general:
>   log_level: INFO
> generate:
>   chunk_word_count: 1000
>   model: merlinite-7b-lab-Q4_K_M
>   num_cpus: 10
>   num_instructions: 100
>   output_dir: generated
>   prompt_file: prompt.txt
>   seed_file: seed_tasks.json
>   taxonomy_base: origin/main
>   taxonomy_path: taxonomy
> serve:
>   gpu_layers: -1
>   host_port: 127.0.0.1:8000
>   max_ctx_size: 4096
>   model_path: models/merlinite-7b-lab-Q4_K_M.gguf
> ```
> 
> ```
> $ ilab serve
> $ ilab chat
> Executing chat failed with: Model merlinite-7b-lab-Q4_K_M is not served by the server.
> ```

Yep, this seems expected to me since we changed the defaults. Maybe a note in the release about this?

File: src/instructlab/chat/chat.py
Comment (russellb) [inline src/instructlab/chat/chat.py:487]: This check won't work if we're talking to a remote model endpoint. It's also making assumptions about how the model is named, which may not be a path to the model at all. That's how it works with our built-in llama-cpp, but it doesn't have to be that way.

**Review by russellb — CHANGES_REQUESTED** [review]

File: src/instructlab/chat/chat.py
Comment (nathan-weinberg) [reply src/instructlab/chat/chat.py:487]: I think one of the other points of the PR is forcing the model to be a path, since we could have duplicate model names. 

Can we even talk to a remote model endpoint at this stage? Doesn't it have to be a local file?

Comment (russellb) [reply src/instructlab/chat/chat.py:487]: > I think one of the other points of the PR is forcing the model to be a path, since we could have duplicate model names.
> 
> Can we even talk to a remote model endpoint at this stage? Doesn't it have to be a local file?

It does not.  You can do `ilab chat --endpoint-url https://anything.openai.compatible/v1`.

Here's a suggested alternative:

Add:

```
DEFAULT_MODEL_OLD = "merlinite-7b-lab-Q4_K_M"
```

and then:
```suggestion
        if model == DEFAULT_MODEL_OLD:
            ... friendly message about using an old config that we know is now broken with instructions on how to fix it ...
```

File: src/instructlab/config.py
Comment (russellb) [reply src/instructlab/config.py:24]: or just this and skip the issue ...

```suggestion
DEFAULT_MODEL_PATH = DEFAULT_MODEL = "models/merlinite-7b-lab-Q4_K_M.gguf"
```

File: src/instructlab/chat/chat.py
Comment (cdoern) [reply src/instructlab/chat/chat.py]: Ok, added this

**Review by russellb — APPROVED** [review]
Thanks for all the follow-up work, @cdoern.

I've played with this again with different configs and options and I'm happy with the behavior now.

**Comment by cdoern** [comment]
@hickeyma your comments were addressed, thanks for the reviews! I added that test back and expected it to fail, I also added `DEFAULT_MODEL_OLD` which I think makes it more clear as to why those constants changed.

The most recent push incorporated @leseb's idea to print the served models when we fail. 

Dropping the blocking changes requested since all of the comments have been incorporated, and i want to ensure we have some time with this in before cutting the release :)

**Pull request merged**
